### PR TITLE
Add explicit dense conjugate dot product

### PR DESCRIPTION
**Pull request by upsj**
Right now, the conjugation happens implicitly in compute_dot, but I think it makes more sense to have both normal and conjugate dot product available.

TODO
- [x] Add true complex tests

### CODE REVIEW

**Review by yhmtsai — CHANGES_REQUESTED** [review]
LGTM. I think the x->conjugate_dot(y) = x^H y is more close to the representation?

File: omp/matrix/dense_kernels.cpp [hunk 1]
```diff
@@ -220,14 +220,35 @@ void compute_dot(std::shared_ptr<const OmpExecutor> exec,
...
+                      matrix::Dense<ValueType> *result)
+{
+#pragma omp parallel for
+    for (size_type j = 0; j < x->get_size()[1]; ++j) {
+        result->at(0, j) = zero<ValueType>();
+    }
+#pragma omp parallel for
+    for (size_type j = 0; j < x->get_size()[1]; ++j) {
+        for (size_type i = 0; i < x->get_size()[0]; ++i) {
+            result->at(0, j) += x->at(i, j) * conj(y->at(i, j));
```
[the remark this answers is not in the document]
Comment (upsj) [reply omp/matrix/dense_kernels.cpp:244]: good point, no idea how I missed that

**Review by pratikvn — APPROVED** [review]
LGTM!

**Comment by upsj** [comment]
I can confirm that this works and passes the tests on CUDA and HIP using the `cuda:101-gnu8-llvm7-intel2019` image.

**Review by yhmtsai — APPROVED** [review]
LGTM

**Pull request merged**